### PR TITLE
Some OIDC refactorings

### PR DESCRIPTION
**Pull request by bnjbvr**
This adds lots of tests for OIDC by abstracting the backend into a trait, and implementing a dummy backend for test purposes.

Also does a few small refactorings / renamings here and there. Best reviewed commit by commit.

### CODE REVIEW

File: crates/matrix-sdk/src/oidc/impl/mod.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+#[cfg(test)]
+pub(crate) use test::*;
+
+pub(super) struct RefreshedSessionTokens {
+    pub access_token: String,
+    pub refresh_token: Option<String>,
+}
+
+#[async_trait::async_trait]
+pub(super) trait OidcImpl: std::fmt::Debug + Send + Sync {
```
Comment (jplatte) [inline crates/matrix-sdk/src/oidc/impl/mod.rs:47]: Shouldn't it be named `OidcBackend` as that's what corresponding bindings are called?

File: crates/matrix-sdk/src/oidc/impl/mod.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
```
Comment (jplatte) [inline crates/matrix-sdk/src/oidc/impl/mod.rs:1]: I find the `impl` module that needs `r#` syntax to be referred to pretty ugly. Do we need a sub-module for these 75 LOC at all?

File: crates/matrix-sdk/src/oidc/impl/mod.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+use super::{AuthorizationCode, OidcError, SessionTokens};
+
+mod server;
+pub(crate) use server::*;
+
+#[cfg(test)]
+mod test;
+#[cfg(test)]
+pub(crate) use test::*;
```
Comment (jplatte) [inline crates/matrix-sdk/src/oidc/impl/mod.rs:39]: Can we please not use `*` imports / re-exports?

File: crates/matrix-sdk/src/oidc/impl/test.rs [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+use crate::oidc::{AuthorizationCode, SessionTokens};
+
+pub(crate) const ISSUER_URL: &str = "https://oidc.example.com/issuer";
+pub(crate) const AUTHORIZATION_URL: &str = "https://oidc.example.com/authorization";
+pub(crate) const REVOCATION_URL: &str = "https://oidc.example.com/revocation";
+pub(crate) const TOKEN_URL: &str = "https://oidc.example.com/token";
+pub(crate) const JWKS_URL: &str = "https://oidc.example.com/jwks";
+
+#[derive(Debug)]
+pub(crate) struct TestImpl {
```
Comment (jplatte) [inline crates/matrix-sdk/src/oidc/impl/test.rs:47]: `MockImpl` might be a better name, maybe?

File: crates/matrix-sdk/src/oidc/tests.rs [hunk 5]
```diff
@@ -29,22 +31,24 @@ use crate::{
...
-    RegisteredClientData {
-        credentials: ClientCredentials::None { client_id: CLIENT_ID.to_owned() },
-        metadata: ClientMetadata {
-            redirect_uris: Some(vec![]), // empty vector is ok lol
-            token_endpoint_auth_method: Some(OAuthClientAuthenticationMethod::None),
-            ..ClientMetadata::default()
-        }
-        .validate()
-        .expect("validate client metadata"),
+pub fn fake_client_credentials() -> ClientCredentials {
```
Comment (jplatte) [inline crates/matrix-sdk/src/oidc/tests.rs:34]: I know we've used this "fake" naming before, but why not just "mock" here as well?

**Review by jplatte — COMMENTED** [review]
Probably too hard to untangle now, but I would have appreciated separate PRs for the test changes and the other refactorings.

**Comment by bnjbvr** [comment]
(Splitting this PR in two parts)

File: crates/matrix-sdk/src/oidc/impl/mod.rs (hunk 1 above)
Comment (bnjbvr) [reply crates/matrix-sdk/src/oidc/impl/mod.rs:47]: Sure

File: crates/matrix-sdk/src/oidc/impl/mod.rs (hunk 2 above)
Comment (bnjbvr) [reply crates/matrix-sdk/src/oidc/impl/mod.rs:1]: Renamed backend/mod.rs. Navigating is pretty fast with LSP, so I'm not too worried about proliferation of files (also the `mod.rs` of OIDC is 1500 lines long, so any attempt to reduce its size is good, IMO)

Comment (jplatte) [reply crates/matrix-sdk/src/oidc/impl/mod.rs:1]: Yeah, I wasn't really concerned about nesting, more about `r#impl`.

**Pull request closed without merging**